File: apps/backend/src/rhesis/backend/app/schemas/metric_tuning.py
"""API shapes for metric tuning cases and the reviews of what a metric said.

A tuning case is one situation a metric has to get right: an input, the answer
being judged and -- where the metric needs one -- a reference answer. It records
no expected verdict. Nobody can honestly say what number a numeric metric
*should* have returned, while saying whether ``0.2`` is wrong is easy, so the
judgement happens after a run instead: a reviewer accepts what the metric said or
rejects it with a comment (domain.local/adr/0005).

The three case fields are the **case payload** -- what the metric is shown --
stored together in ``prompt.content`` because a tuning case puts the metric in
the system-under-test role (ADR-0003). Reviews are stored apart, in
``test.test_metadata``, and are never shown to the metric.

``outcome`` and ``review`` are both derived on read. A review is only still
standing while the metric's verdict has not materially changed under the metric's
current threshold, which is a question that cannot be answered once and stored.
"""

from datetime import datetime
from enum import Enum
from typing import List, Optional, Union

from pydantic import UUID4, BaseModel, ConfigDict, Field

from rhesis.backend.app.schemas import Base
from rhesis.backend.app.schemas.metric_tuning_metadata import ReviewDecision, TuningRunStatus
from rhesis.backend.app.schemas.metric_types import ScoreType, ThresholdOperator


class TuningCaseOutcome(str, Enum):
    """How one case stands after the latest run.

    ``UNREVIEWED`` is never counted as accepted: a set nobody looked at must not
    report itself as perfect, and the cases an edit just broke must not read as
    successes until someone looks.
    """

    ACCEPTED = "accepted"
    REJECTED = "rejected"
    # The metric call failed. Not the same as a verdict a reviewer rejected.
    ERRORED = "errored"
    UNREVIEWED = "unreviewed"


class UnreviewedReason(str, Enum):
    """Why an unreviewed case is unreviewed. The work is the same either way,
    but a case that lost a review says so rather than looking untouched."""

    NEVER_JUDGED = "never_judged"
    INVALIDATED = "invalidated"


class MetricTuningCaseBase(Base):
    # The input given to the system under test.
    input: str
    # The answer the metric has to judge.
    output: str
    # What the system under test should have answered, for a metric that judges
    # against a reference. Never an "expected" anything -- that word belonged to
    # the retired expected verdict and the two were indistinguishable in the grid.
    reference_answer: Optional[str] = None


class MetricTuningCaseCreate(MetricTuningCaseBase):
    """A new case. The input and the answer being judged are both required --
    they are what the metric is run over, so without them there is nothing to
    run."""


class MetricTuningCaseUpdate(Base):
    """Partial update -- only the fields present are applied."""

    input: Optional[str] = None
    output: Optional[str] = None
    reference_answer: Optional[str] = None


class MetricTuningCaseResult(BaseModel):
    """What the metric said about this case in the latest run.

    Absent until the metric has been run over the case. ``error`` set means the
    call failed, which is deliberately not the same as a verdict a reviewer would
    reject.

    Plain ``BaseModel``, not the shared ``Base``: this is a value on a case, not a
    row. ``Base`` would add ``id``/``nano_id``/``project_id``, and an ``id`` here
    would advertise a handle that does not exist.
    """

    # The metric's own verdict, as a string, whatever its score type.
    verdict: Optional[str] = None
    # The metric's explanation, where it produces one.
    reasoning: Optional[str] = None
    # Set when the metric call failed for this case.
    error: Optional[str] = None
    evaluated_at: Optional[str] = None


class MetricTuningReview(BaseModel):
    """The review that currently stands for a case.

    ``verdict`` is the raw verdict the reviewer was judging, so the interface can
    show what the judgement was about. The review history behind this is not
    exposed: what a reader needs is the judgement that holds now and the comment
    that came with it.
    """

    decision: ReviewDecision
    # What is wrong with the verdict. Always present on a rejection.
    comment: Optional[str] = None
    verdict: Optional[str] = None
    reviewed_at: Optional[str] = None


class MetricTuningReviewCreate(BaseModel):
    """A reviewer's judgement of the verdict a case currently carries.

    The verdict being judged is read server-side from the stored result rather
    than sent, so a review can never claim to be about something the metric did
    not say.
    """

    decision: ReviewDecision
    # Required on a rejection: the comment is what someone reads when rewriting
    # the evaluation prompt, so a rejection without one records nothing useful.
    comment: Optional[str] = None


class MetricTuningCase(MetricTuningCaseBase):
    id: UUID4
    # The latest run's result for this case, or None if it has never been run.
    result: Optional[MetricTuningCaseResult] = None
    # Derived on read, both of them -- see the module docstring.
    outcome: TuningCaseOutcome = TuningCaseOutcome.UNREVIEWED
    review: Optional[MetricTuningReview] = None
    unreviewed_reason: Optional[UnreviewedReason] = None
    created_at: Union[datetime, str]
    updated_at: Union[datetime, str]

    model_config = ConfigDict(from_attributes=True)


class TuningAgreement(BaseModel):
    """How much of what the metric said the reviewer accepted.

    This is the one number an author watches while editing an evaluation prompt:
    change the wording, run again, see whether it went up.

    ``ratio`` is ``None`` when nothing has been judged, never ``1.0`` -- a set
    nobody has looked at has no agreement rather than a perfect one. It is
    computed from the stored reviews on every read and never written down, so a
    review a run has just invalidated stops counting immediately.

    ``judged`` travels with the ratio because a ratio without its denominator is
    not a measurement: three out of three should not read like a solved problem.
    """

    # accepted / (accepted + rejected). None when nothing has been judged.
    ratio: Optional[float] = None
    # The denominator -- accepted plus rejected, and nothing else.
    judged: int = 0
    accepted: int = 0
    rejected: int = 0
    # Left out of the ratio and reported beside it, never counted as accepted.
    unreviewed: int = 0
    # The metric call failed. Left out too, and kept apart from the verdicts so a
    # flaky provider never reads as a bad metric.
    errored: int = 0


class MetricTuningRun(BaseModel):
    """The state of a metric's latest tuning run.

    Only the latest is kept -- a new run overwrites the previous one, per
    domain.local/adr/0004. A metric that has never been run reads as
    ``never_run`` rather than as an absent object, so the interface has one
    shape to render either way.

    Plain ``BaseModel``, not the shared ``Base``: a run is stored in JSONB and is
    deliberately not a row (ADR-0004), so it has no id to expose. Inheriting
    ``Base`` would put a null ``id`` on every response and suggest otherwise.
    """

    status: TuningRunStatus = TuningRunStatus.NEVER_RUN
    started_at: Optional[str] = None
    completed_at: Optional[str] = None
    # How many cases the run set out to cover, and how many it has finished.
    # The pair is what makes progress reportable while a run is still going.
    total_cases: int = 0
    completed_cases: int = 0
    # Cases whose metric call failed. Reported apart from the verdicts.
    errored_cases: int = 0
    # Why the run as a whole failed. A single case failing does not fail a run.
    error: Optional[str] = None
    # Not part of the run at all -- reviews are written between runs and change
    # this without one. It travels with the run because that is the one thing the
    # tab already re-reads while a run is going, and the number has to move as the
    # cases land. Derived on every read; see ``TuningAgreement``.
    agreement: TuningAgreement = TuningAgreement()
    # True when the metric has changed in a verdict-affecting way since this run
    # started, so its agreement belongs to the earlier metric. Derived from the
    # stored fingerprint on every read; a run without one answers False, because
    # unknown is not the same as out of date.
    predates_metric: bool = False


class ImprovedMetricFields(BaseModel):
    """The metric fields a model may rewrite from a reviewer's rejections.

    Deliberately narrow: the evaluation fields and nothing else. No ids, no
    relations, no ``metric_scope`` -- a rejection says the metric judged one case
    wrongly, which is not an opinion about which turn shapes it applies to.

    Every field is required but nullable rather than optional-with-a-default, so
    the JSON schema handed to the model lists them all: a provider running strict
    structured output rejects a schema whose properties are not all required, and
    a numeric metric genuinely has no ``categories`` to give.

    ``score_type`` and ``categories`` are here because the model has to reason
    about the score bands coherently, not because it may move them -- both are
    overwritten with the metric's current values before this is returned. An
    improvement that changed ``score_type`` would invalidate every review for the
    metric, which is not something a button does quietly (domain.local/adr/0006).
    """

    name: str = Field(description="Title Case with spaces, e.g. 'Factual Accuracy'")
    description: str
    evaluation_prompt: str = Field(description="The evaluation criteria, no template placeholders")
    evaluation_steps: str
    reasoning: str
    explanation: str
    score_type: ScoreType
    min_score: Optional[float]
    max_score: Optional[float]
    threshold: Optional[float]
    threshold_operator: Optional[ThresholdOperator]
    categories: Optional[List[str]]
    passing_categories: Optional[List[str]]


class MetricTuningImprovement(BaseModel):
    """A proposed rewrite of a metric, read off the rejections its reviewers wrote.

    Producing one never writes anything. The reviewer sees the current fields
    beside these and applies them with an ordinary metric update, or does not --
    an in-place rewrite of the evaluation prompt the reviews were made against
    would have no diff and no undo (ADR-0006).

    ``changed`` is what the dialog shows; the fields it leaves out are named as
    unchanged rather than hidden, so the reviewer can see the rewrite left them
    alone. ``rejections_used`` is the header: how many comments this was written
    from.
    """

    improvement: ImprovedMetricFields
    # Names of the fields whose proposed value differs from the metric's current one.
    changed: List[str] = []
    rejections_used: int = 0


__all__ = [
    "ImprovedMetricFields",
    "MetricTuningCase",
    "MetricTuningCaseCreate",
    "MetricTuningCaseResult",
    "MetricTuningCaseUpdate",
    "MetricTuningImprovement",
    "MetricTuningReview",
    "MetricTuningReviewCreate",
    "MetricTuningRun",
    "TuningAgreement",
    "ReviewDecision",
    "TuningCaseOutcome",
    "UnreviewedReason",
]

File: apps/frontend/src/utils/api-client/interfaces/metric-tuning.ts
import { UUID } from 'crypto';
import type { ScoreType, ThresholdOperator } from './metric';

/**
 * What the metric said about a case in the latest run.
 *
 * `error` set means the call failed, which is deliberately not the same thing
 * as the metric being wrong.
 */
export interface MetricTuningCaseResult {
  /** The metric's own verdict, as a string, whatever its score type. */
  verdict: string | null;
  /** The metric's explanation, where it produces one. */
  reasoning: string | null;
  /** Why the metric call failed for this case, when it did. */
  error: string | null;
  evaluated_at: string | null;
}

/** What a reviewer said about the verdict a metric gave. */
export type TuningDecision = 'accepted' | 'rejected';

/** The annotation that currently stands for a case. */
export interface MetricTuningAnnotation {
  id: UUID;
  decision: TuningDecision;
  /** Required on a rejection, always null on an accept. */
  comment: string | null;
  /** The raw metric verdict this judgement was made against. */
  verdict: string | null;
  annotated_at: string | null;
}

export type MetricTuningOutcome =
  | 'accepted'
  | 'rejected'
  | 'errored'
  | 'unannotated';

/**
 * A metric tuning case: one input plus the answer the metric has to judge.
 *
 * The case records no expected verdict — nothing is compared for equality.
 * After a run, a reviewer reads what the metric said and either accepts it or
 * rejects it with a comment, which is what `annotation` and `outcome` carry.
 */
export interface MetricTuningCase {
  id: UUID;
  /** The input given to the system under test. */
  input: string;
  /** The answer the metric has to judge. */
  output: string;
  /** What the system under test should have answered, when the metric needs one. */
  reference_answer: string | null;
  /** The latest run's result, or null if the metric has not been run over it. */
  result: MetricTuningCaseResult | null;
  outcome: MetricTuningOutcome;
  /** The standing judgement, or null when the case is unannotated/errored. */
  annotation: MetricTuningAnnotation | null;
  /** Why an unannotated case is unannotated; null otherwise. */
  unannotated_reason: 'never_judged' | 'invalidated' | null;
  created_at: string;
  updated_at: string;
}

/**
 * Where a metric's latest tuning run got to.
 *
 * `never_run` rather than an absent object, so the tab renders one shape either
 * way.
 */
export type TuningRunStatus = 'never_run' | 'running' | 'completed' | 'failed';

/**
 * How much of what the metric said the reviewer accepted — the one number to
 * watch while editing an evaluation prompt.
 *
 * `ratio` is null when nothing has been judged, never 1 — a set nobody has
 * looked at has no agreement rather than a perfect one. Unannotated and errored
 * cases are counted out of it and reported beside it instead.
 */
export interface MetricTuningAgreement {
  /** accepted / (accepted + rejected), or null when nothing has been judged. */
  ratio: number | null;
  /** The denominator. Never show the ratio without it. */
  judged: number;
  accepted: number;
  rejected: number;
  /** Left out of the ratio, never counted as accepted. */
  unannotated: number;
  /** The metric call failed — left out too, and reported apart. */
  errored: number;
  /**
   * Test results where someone overruled this metric on a real run. Outside the
   * ratio on purpose: that measures the curated tuning set, and a metric with no
   * tuning cases at all can still have a count here.
   */
  disagreements_in_runs: number;
}

/**
 * A metric's latest tuning run. Only the latest is kept — a new run overwrites
 * the previous one.
 */
export interface MetricTuningRun {
  status: TuningRunStatus;
  started_at: string | null;
  completed_at: string | null;
  /** How many cases the run covers, and how many it has finished so far. */
  total_cases: number;
  completed_cases: number;
  /** Cases whose metric call failed. Counted apart from the verdicts. */
  errored_cases: number;
  /** Why the run as a whole failed. One case failing does not fail a run. */
  error: string | null;
  /**
   * Recomputed from the stored annotations on every read, so a judgement
   * recorded between runs moves it without a run.
   */
  agreement: MetricTuningAgreement;
  /**
   * True when the metric has changed in a verdict-affecting way since this run
   * started — so its agreement belongs to the earlier metric, not the one on
   * screen. Renaming a metric does not set this; editing its evaluation prompt
   * does.
   */
  predates_metric: boolean;
}

/**
 * The metric fields a model may rewrite from a reviewer's rejections.
 *
 * `score_type` and `categories` always come back as the metric already has
 * them — an improvement that moved either would invalidate every annotation on
 * the metric.
 */
export interface ImprovedMetricFields {
  name: string;
  description: string;
  evaluation_prompt: string;
  evaluation_steps: string;
  reasoning: string;
  explanation: string;
  score_type: ScoreType;
  min_score: number | null;
  max_score: number | null;
  threshold: number | null;
  threshold_operator: ThresholdOperator | null;
  categories: string[] | null;
  passing_categories: string[] | null;
}

/**
 * A proposed rewrite of a metric, read off the rejections its reviewers wrote.
 *
 * Producing one saves nothing. The reviewer sees the current fields beside these
 * and applies them with an ordinary metric update, or closes the dialog.
 */
export interface MetricTuningImprovement {
  improvement: ImprovedMetricFields;
  /** Fields whose proposed value differs from the metric's current one. */
  changed: (keyof ImprovedMetricFields)[];
  /** How many rejections it was written from, across all three sources. */
  rejections_used: number;
  /** Tuning cases someone curated for judging this metric. */
  tuning_rejections_used: number;
  /** People overruling this metric on a real test result. */
  run_rejections_used: number;
  /**
   * How many of those exist. Higher than `run_rejections_used` when the cap
   * bit, so the difference is shown rather than silently dropped.
   */
  run_rejections_found: number;
  /** Explorer tests labelled against what this metric said about them. */
  explorer_rejections_used: number;
  explorer_rejections_found: number;
}

export interface MetricTuningCaseCreate {
  input: string;
  output: string;
  reference_answer?: string | null;
}

/** Partial update — only the fields present are applied. */
export interface MetricTuningCaseUpdate {
  input?: string;
  output?: string;
  reference_answer?: string | null;
}

/** A reviewer's judgement of what the metric said. */
export interface MetricTuningAnnotationCreate {
  decision: TuningDecision;
  /** Required on a rejection — the API rejects a blank one. */
  comment?: string | null;
}

export interface MetricTuningCaseDeleteResponse {
  deleted: boolean;
  case_id: string;
}
